### PR TITLE
issue_194_PR

### LINKED ISSUE #194: Add Application Function Id (afId) or Sponsor Id
**Problem description**
 Service Provider would like to track QoD session data usage for third party usage reporting or monetization purposes. 

**Possible evolution**
 Add Sponsor Id or Application Function ID (afID) in QoD API. These IDs are defined in 3GPP specs TS 29.122 (SCEF)  and TS 29.522 (NEF).  This will enable QoD session usage tracking based on application function ID in the network.

### PR DESCRIPTION
**Pull request by SyeddR**
#### What type of PR is this?
* PR for introducing application consumer Id as discussed in #194 
Add one of the following kinds:

* enhancement/feature

#### What this PR does / why we need it:

* Identifies the application consumer that initiates a qos session. The ID is systematically relayed to downstream systems, including but not limited to, the Network Exposure Function (NEF) and the Policy Control Function (PCF) within the 5G framework. The primary utility of this transmission is to facilitate the application of relevant policies, ensuring that the application consumer is appropriately tagged within usage records.
     

#### Which issue(s) this PR fixes:

Fixes #194 

#### Special notes for reviewers:

For lack of the better term i am proposing "applicationConsumerId" as string type field. Open for suggestions

#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.

```
docs

```

### CODE REVIEW

**Comment by eric-murray** [comment]
Hi @SyeddR

I had understood that the `applicationConsumerId` must be associated with one, and only one application. Is that the case? If so, my preference would be for this to be a sub-property of the `applicationServer` property, the reason being that I'm not a fan of "flat" input parameter structures, and much prefer that related parameters be grouped, particularly when some (such as this one) are optional.

Of course, you could argue that this should also apply to `applicationServerPorts`, and I'd agree with you :-)

I also agree that adding this property to the `ApplicationServer` schema requires some thought, as we need to ensure that the API consumer provides at least one IP address, but there are ways in OAS that that can be mandated whilst still allowing `applicationConsumerId` to be optionally specified.

**Comment by SyeddR** [comment]
> Hi @SyeddR
> 
> I had understood that the `applicationConsumerId` must be associated with one, and only one application. Is that the case? If so, my preference would be for this to be a sub-property of the `applicationServer` property, the reason being that I'm not a fan of "flat" input parameter structures, and much prefer that related parameters be grouped, particularly when some (such as this one) are optional.
> 
> Of course, you could argue that this should also apply to `applicationServerPorts`, and I'd agree with you :-)
> 
> I also agree that adding this property to the `ApplicationServer` schema requires some thought, as we need to ensure that the API consumer provides at least one IP address, but there are ways in OAS that that can be mandated whilst still allowing `applicationConsumerId` to be optionally specified.

**Comment by jlurien** [comment]
I suggested to open an issue on Identity and Consent WG about this topic, as identification of Application is being discussed there in the context of authentication, and access token should contain all required information.

File: code/API_definitions/qod-api.yaml [hunk 1]
```diff
@@ -433,6 +433,7 @@ components:
       description: Attributes required to create a session
       type: object
       properties:
+
```
Comment (eric-murray) [inline code/API_definitions/qod-api.yaml:436]: Why the blank line here?

Comment (SyeddR) [reply code/API_definitions/qod-api.yaml]: Fixed it.

**Comment by eric-murray** [comment]
@SyeddR 
I'm fine with the proposal, but updated the `ApplicationServer` schema description to clarify that `applicationId` itself is not sufficient to identify the application server, and that an IP address is still required.

I'm a bit concerned that the schema definition will not catch API consumers who only provide the `applicatonId` (because that will satisfy the `minProperties: 1` directive), so am wondering if we need to re-organise the schema so that a single `ipAddress` property is now `required`. One for discussion.

There appear to be some conflicts now with the main branch (which maybe I introduced - apologies if that is the case), Can you synchronise your fork with main to see if that fixes things?

**Comment by SyeddR** [comment]
> There appear to be some conflicts now with the main branch (which maybe I introduced - apologies if that is the case), Can you synchronise your fork with main to see if that fixes things?

@eric-murray My fork shows all synced with latest updates. Can you try to resolve the conflict at your end.

**Review by eric-murray — APPROVED** [review]

**Comment by eric-murray** [comment]
@SyeddR
OK, fixed and now approved from my side. It seems my original edits were not properly merged into your own fork, but subsequent edits were, so all good now.

File: code/API_definitions/qod-api.yaml [hunk 2]
```diff
@@ -858,14 +858,21 @@ components:
...
         * `ipv6Address`
+
+        In addition, the developer can specify an `applicationId` for policy control / audit purposes. The `applicationId` is not sufficient in itself to identity the application server.
       type: object
       properties:
+        applicationId:
+          description: |
+            A public identifier that uniquely identifies the application initiating a QoS session. The applicationId is more granular than the clientId, which identifies the unique customer. A client or customer could have one or more applications that need to be tracked. The Id is systematically relayed to downstream systems, including but not limited to, the Network Exposure Function (NEF) and the Policy Control Function (PCF) within the 5G framework.
+            The primary utility of this transmission is to facilitate the application of relevant policies, ensuring that the application is appropriately tagged within usage records.
+          type: string
```
Comment (RandyLevensalor) [inline code/API_definitions/qod-api.yaml:874]: Could we some constraints on the string size and values?
minLength: 3
maxLength: 256 (default value)
pattern: (should we exclude strings like "??? or ***"

**Review by RandyLevensalor — COMMENTED** [review]
Could we apply some constraints to this?

**Comment by hdamker** [comment]
Converted to draft status as agreed within our call on Nov 3rd.

File: code/API_definitions/qod-api.yaml (hunk 2 above)
Comment (RandyLevensalor) [reply code/API_definitions/qod-api.yaml:874]: @SyeddR any response to this?

**Comment by hdamker** [comment]
This PR is stale and can't be merged into the current (splitted) API structure anymore.

**Pull request closed without merging**